gh-130425: Add "Did you mean [...]" suggestions for `del obj.attr`

Pranjal095 opened the pull request:
This pull request replaces the earlier PR [#130799](https://github.com/python/cpython/pull/130799) (now closed) and carries over that set of commits, preserving original authorship and history. An extra commit has been included to address a linting issue that was present in the previous version.

Issue: [#130425](https://github.com/python/cpython/issues/130425)

<!-- gh-issue-number: gh-130425 -->
* Issue: gh-130425
<!-- /gh-issue-number -->

sobolevn committed:
gh-130425: Add "Did you mean" suggestion for `del obj.attr`

sobolevn committed:
Update NEWS

Pranjal095 committed:
gh-130428: Add tests for delattr suggestions

Pranjal095 committed:
Refactored getattr and delattr tests

Pranjal095 committed:
Added else branches to handle unrecognized operations

Pranjal095 committed:
Refactor getattr and setattr suggestion tests

nedbat committed:
docs: -I also implies -P (#131539)

sobolevn committed:
GH-134236: make regen-all (GH-134237)

tomasr8 committed:
GH-131798: Optimize away isinstance calls in the JIT (GH-134369)

Pranjal095 committed:
Refactor getattr and delattr suggestion tests

Pranjal095 committed:
Fix linting issue

picnixz commented on Lib/test/test_traceback.py:
```diff
@@ -4074,63 +4077,88 @@ class CaseChangeOverSubstitution:
+        obj = A()
+        self.assertIn("'bluch'", self.get_suggestion(obj, 'blach'))
+        self.assertIn("'bluch'", self.get_suggestion(obj, '_luch'))
+        self.assertIn("'bluch'", self.get_suggestion(obj, '_bluch'))
```
Let's not change the tests here. I think there was a discussion where we changed `obj = A()` but to reduce the diff, we can avoid this.

picnixz commented on Lib/test/test_traceback.py:
```diff
@@ -4074,63 +4077,88 @@ class CaseChangeOverSubstitution:
-        self.assertIn("'_bluch'", self.get_suggestion(partial(B().method, '_blach')))
-        self.assertIn("'_bluch'", self.get_suggestion(partial(B().method, '_luch')))
-        self.assertIn("'_bluch'", self.get_suggestion(partial(B().method, 'bluch')))
+        if hasattr(self, 'test_with_method_call'):
```
I don't understand why we have this if. If this is only specific to some subclass, add a specific test there.

picnixz commented on Lib/test/test_traceback.py:
```diff
@@ -4074,63 +4077,88 @@ class CaseChangeOverSubstitution:
         for name in ("b", "v", "m", "py"):
             with self.subTest(name=name):
-                actual = self.get_suggestion(MyClass, name)
+                actual = self.get_suggestion(obj, name)
```
```suggestion
                actual = self.get_suggestion(MyClass(), name)
```

picnixz commented on Lib/test/test_traceback.py:
```diff
@@ -4034,7 +4033,11 @@ def callable():
+    """
+    Subclasses need to implement the get_suggestion method.
+    """
+    def test_suggestions(self):
```
```suggestion
    """

    def test_suggestions(self):
```

picnixz commented on Lib/test/test_traceback.py:
```diff
@@ -4020,8 +4020,7 @@ def test_dont_swallow_subexceptions_of_falsey_exceptiongroup(self):
 
-
-class SuggestionFormattingTestBase:
+class SuggestionFormattingTestBaseParent:
```
The class name I suggested wasn't the best actually... so maybe you should find a better one. And please keep 2 blank lines between classes.

Pranjal095 committed:
Add requested changes

Pranjal095 committed:
Merge remote-tracking branch 'upstream/main' into delattr-suggestions-re

picnixz commented on Lib/test/test_traceback.py:
```diff
@@ -4171,13 +4187,36 @@ def __getattr__(self, attr):
+
+        self.assertIn("'_bluch'", self.get_suggestion(partial(B().method, '_blach')))
+        self.assertIn("'_bluch'", self.get_suggestion(partial(B().method, '_luch')))
+        self.assertIn("'_bluch'", self.get_suggestion(partial(B().method, 'bluch')))
```
```suggestion
        self.assertIn("'_bluch'", self.get_suggestion(partial(B().method, '_blach')))
        self.assertIn("'_bluch'", self.get_suggestion(partial(B().method, '_luch')))
        self.assertIn("'_bluch'", self.get_suggestion(partial(B().method, 'bluch')))
```

Since this is new code, you can do `obj = B()`.

picnixz commented on Misc/NEWS.d/next/Core_and_Builtins/2025-02-22-01-23-23.gh-issue-130425.x5SNQ8.rst:
```diff
@@ -0,0 +1,2 @@
+Add ``"Did you mean: 'attr'?"`` suggestion when using ``del obj.attr`` if ``attr``
```
Let's also add a `What's New` entry as this is a user-facing change. For that you need to update `Doc/whatsnew/3.15.rst`. I don't know if there is a built-in section or language change section already. If not, look at 3.14.rst where we indicated that we improved error messages.

picnixz commented on Objects/dictobject.c:
```diff
@@ -6939,6 +6939,7 @@ store_instance_attr_lock_held(PyObject *obj, PyDictValues *values,
         PyErr_Format(PyExc_AttributeError,
                         "'%.100s' object has no attribute '%U'",
                         Py_TYPE(obj)->tp_name, name);
+        _PyObject_SetAttributeErrorContext(obj, name);
```
```suggestion
        (void)_PyObject_SetAttributeErrorContext(obj, name);
```

Since we don't use the return value.

Pranjal095 committed:
Add requested changes

picnixz commented on Doc/whatsnew/3.15.rst:
```diff
@@ -175,6 +175,20 @@ Other language changes
 * Several error messages incorrectly using the term "argument" have been corrected.
   (Contributed by Stan Ulbrych in :gh:`133382`.)
 
+* The interpreter now provides helpful suggestions when :func:`delattr` fails due to a missing attribute. When an attribute name that closely resembles an existing attribute is used, the interpreter will suggest the correct attribute name in the error message. For example:
```
Can you wrap the text under 80 characters per line?

picnixz commented on Doc/whatsnew/3.15.rst:
```diff
@@ -175,6 +175,20 @@ Other language changes
+   ...     pass
+   >>> a = A()
+   >>> a.abcde = 1
+   >>> del a.abcdf
```
I think you need `# doctest: +ELLIPSIS`

picnixz commented on Doc/whatsnew/3.15.rst:
```diff
@@ -175,6 +175,20 @@ Other language changes
+     ...
+   AttributeError: 'A' object has no attribute 'abcdf'. Did you mean: 'abcde'?
+
+(Contributed by [sobolevn] and Pranjal Prajapati in :gh:`136588`.)
```
soblevn is Nikita Sobolev

Pranjal095 committed:
Add requested changes

Pranjal095 committed:
Fix doc indentation issue

StanFromIreland commented on Doc/whatsnew/3.15.rst:
```diff
@@ -175,6 +175,24 @@ Other language changes
 * Several error messages incorrectly using the term "argument" have been corrected.
   (Contributed by Stan Ulbrych in :gh:`133382`.)
 
+* The interpreter now provides helpful suggestions
```
```suggestion
* The interpreter now provides a suggestion
```
It is limited to one suggestion, no? I also think "helpful" is not something we can state.

Pranjal095 committed:
Make requested change

picnixz commented on Doc/whatsnew/3.15.rst:
```diff
@@ -175,6 +175,24 @@ Other language changes
+  when :func:`delattr` fails due to a missing attribute.
+  When an attribute name that closely resembles an existing attribute is used,
+  the interpreter will suggest the correct attribute
+  name in the error message. For example:
```
```suggestion
* The interpreter now tries to provide a suggestion when
  :func:`delattr` fails due to a missing attribute. For example:
```

picnixz commented on Lib/test/test_traceback.py:
```diff
@@ -4655,6 +4695,49 @@ class CPythonSuggestionFormattingTests(
+    unittest.TestCase,
+):
+    """
+    Same set of tests (for attribute access) as above using the pure Python implementation of
```
Wrap the docstring to 80 chars.

Pranjal095 committed:
Make requested changes

picnixz commented on Doc/whatsnew/3.15.rst:
```diff
@@ -175,6 +175,24 @@ Other language changes
   (Contributed by Stan Ulbrych in :gh:`133382`.)
 
+* The interpreter now provides a suggestion tries to provide a suggestion when
+  :func:`delattr` fails due to a missing attribute.
```
```suggestion
* The interpreter now tries to provide a suggestion when
  :func:`delattr` fails due to a missing attribute.
```

picnixz commented on Doc/whatsnew/3.15.rst:
```diff
@@ -175,6 +175,24 @@ Other language changes
+  :func:`delattr` fails due to a missing attribute.
+  When an attribute name that closely resembles an existing attribute is used,
+  the interpreter will suggest the correct attribute
+  name in the error message. For example:
```
```suggestion
  When an attribute name that closely resembles an existing attribute is used,
  the interpreter will suggest the correct attribute name in the error message.
  For example:
```

picnixz commented on Doc/whatsnew/3.15.rst:
```diff
@@ -175,6 +175,24 @@ Other language changes
+
+  .. doctest::
+
+    >>> class A:
```
Align `>` and the following lines with the `d` of doctest.

picnixz commented on Lib/test/test_traceback.py:
```diff
@@ -4034,7 +4034,13 @@ def callable():
+
+class BaseSuggestionTests(SuggestionFormattingTestMixin):
+    """
+    Subclasses need to implement the get_suggestion method.
```
This docstring doesn't really help so either remove it or improve it.

Pranjal095 committed:
Add requested changes

picnixz reviewed (approved):
This looks good now. 

I'll let @sobolevn have a look as well as he's credited as a contributor as well.

Pranjal095 committed:
Merge branch 'main' into delattr-suggestions-re

encukou committed:
Parametrize getattr/setattr in get_suggestion

This allows moving test_suggestions_with_method_call back to
test_getattr_suggestions_underscored, so that delattr is tested
for these as well.

encukou committed:
Run test_suggestions_for_same_name for delattr too

Pranjal095 committed:
Merge pull request #1 from encukou/delattr-suggestions-re

Delattr suggestions re

encukou reviewed (approved):
LGTM

@sobolevn, I plan to merge around Friday if there are no objections.
